File: Level-2/counting_sort.c
/*
 * Date: 2017-11-29
 *
 * Description:
 * Implement counting sort.
 *
 * Approach:
 * This works for integer only and sorts in O(n) if we know the range of input
 * elements is from 0 to k.
 *
 * Problem:
 * As k increases and dominates over n, both time and space complexity
 * increases. Like if k = n^2 then complexity of this algorithm becomes O(n^2).
 * To solve this problem radix sort is used.
 *
 * Complexity:
 * O(n + k)
 */

#include "stdio.h"
#include "stdlib.h"

void print_array(int arr[], int n, char *msg) {
  int i = 0;
  printf("*********** %s *****************\n", msg);
  for (i = 0; i < n; i++)
    printf("%d ", arr[i]);
  printf("\n\n");
}

int main() {
  int i = 0, j = 0, l = 0;
  int n = 0;
  int *a = NULL, *count = NULL;
  int k = 0;

  printf("Enter range of elements: ");
  scanf("%d", &k);
  k += 1;  // +1 to accommodate 0 to k.
  count = (int *)malloc(sizeof(int) * k);

  // Set count to 0 for each element.
  for (i = 0; i < k; i++)
    count[i] = 0;

  printf("Enter number of elements: ");
  scanf("%d", &n);
  a = (int *)malloc(sizeof(int) * n);
  for (i = 0; i < n; i++) {
    printf("Enter element[%d]: ", i);
    scanf("%d", &a[i]);

    // Take count of each element.
    count[a[i]]++;
  }
  print_array(a, n, "Inserted Array");

  // Update array as per counts of each element.
  for (i = 0; i < k; i++) {
    for (j = 0; j < count[i]; j++)
      a[l++] = i;
  }
  print_array(a, n, "Sorted Array");
  return 0;
}

File: Level-2/counting_sort.py
# Date: 2020-08-30
#
# Description:
# Implement counting sort.
#
# Approach:
# This works for integer only and sorts in O(n). We has to find the range(min
# to max - k) of elements and allocate an array of that size. Then keep of
# saving the counts of each element that appears.
#
# Problem:
# As range of elements(k) increases and dominates over n, both time and space
# complexity increases. Like if k = n^2 then complexity of this algorithm
# becomes O(n^2).
# In such cases radix sort can be used.
#
# Reference:
# https://www.hackerearth.com/practice/algorithms/sorting/counting-sort/tutorial/
#
# Complexity:
# Time - O(n + k)
# Space - O(k)

def counting_sort(A):
    if not A:
        return A
    _min = min(A)
    _max = max(A)

    counter = [0 for _ in range(_max - _min + 1)]

    for n in A:
        counter[n - _min] += 1

    i = 0
    for value, count in enumerate(counter):
        for _ in range(count):
            A[i] = value + _min
            i += 1

def sort(A):
    counting_sort(A)
    return A

assert sort([3, 4, 5, 2, 1]) == [1, 2, 3, 4, 5]
assert sort([3, 4, 5, 2, 1, 6]) == [1, 2, 3, 4, 5, 6]
assert sort([]) == []
assert sort([1]) == [1]
assert sort([2, 1]) == [1, 2]
assert sort([-2, 1]) == [-2, 1]
assert sort([-10, -20, 1, 10, 5, 0, 7, 9]) == [-20, -10, 0, 1, 5, 7, 9, 10]
